### PR TITLE
cpio: changes to make it usable in other u-root tools

### PR DESCRIPTION
**Pull request by rminnich**
Reduce cmds/cpio to a simple program that uses the cpio package.

create the uroot/cpio and uroot/cpio/newc packages.

Rework some tests.

The cpio command continues to work as before, and we can nw create
archives without needing sudo.

It's also now very easy to add new formats, if we care (I doubt we do, but ...)

Missing is uroot/cpio/cpio_test.go, I'll get to that once we have some idea
if this change is sensible.

Questions:
o is the layout of the source file tree right (i.e. uroot/cpio{,/newc})
o given that cpio and newc are dead formats, would it have been
  better to use the function in the cpio that writes newc format
  in build and let that suffice? IOW, is this change more than
  needed given the unimportance of cpio?
o one goal of u-root is that anyone be able to take individual comamnds and use them,
  standalone, in other projects. This is why I've tried to keep uroot/ small and only
  for things specific to u-root. This change violates that rule for cpio.
  Do we care? I'm hoping this is not a trend.

Signed-off-by: Ronald G. Minnich <rminnich@gmail.com>

### CODE REVIEW

**Comment by rjoleary** [comment]
It is a shame that Go forbids you from importing symbols exported from a "main" package. The alternative is for the package to be split into a pair of packages. This was what was done for fmap which I think should work similarly for cpio:

- `cmds/fmap`: contains the main package, parses command lines and creates `io.Reader`s
- `cmds/fmap/lib`: contains the parts of fmap which can be imported to be used by other packages

It is still possible to install fmap using `go get github.com/u-root/u-root/cmds/fmap`.

To answer the questions:
o I don't know what the correct layout should be for nesting packages. Even the standard go packages are inconsistent at times: "encoding","encoding/binary","encoding/xml" vs "cmd/go","cmd/go/internal/vet","cmd/go/internal/tool","cmd/go/internal/list"
o There has been some demand for a cpio reader/writer written in Go. For example: https://github.com/moby/moby/issues/17484 (I've gone crazy or docker got renamed to moby recently).
o Why can't cpio be in the cmds directory like it has been done with fmap?

**Comment by rminnich** [comment]
Ryan, super helpful comments, it's what I tried to do first but screwed up. I'm going to follow your approach.

**Comment by rminnich** [comment]
So, Ryan, the one thing I thin I would change from your suggestion is to use pkg, not lib.

pkg is a name with meaning in go, lib is not.

It's true that the go source is not consistent, I think your fmap idea is the best compromise.

File: cmds/cpio/pkg/newc/newc.go [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+
+func (f *Header) String() string {
+
+	return fmt.Sprintf("Ino %d Mode %#o UID %d GID %d Nlink %d Mtime %#x FileSize %d Major %d Minor %d RMajor %d Rminor %d NameSize %d",
+		f.Ino,
+		f.Mode,
+		f.UID,
+		f.GID,
+		f.Nlink,
+		// what a mess. This fails on travis.
```
Comment (hugelgupf) [inline cmds/cpio/pkg/newc/newc.go]: Why does this fail? :/

File: cmds/cpio/pkg/newc/newc.go [hunk 2]
```diff
@@ -0,0 +1,243 @@
+// Copyright 2013-2017 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// newc implements the interface for new type cpio files.
```
Comment (hugelgupf) [inline cmds/cpio/pkg/newc/newc.go]: suggestion: move to u-root/pkg/cpio/newc

File: cmds/cpio/pkg/newc/newc.go [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+		f.Minor,
+		f.Rmajor,
+		f.Rminor,
+		f.NameSize)
+}
+
+type newcReader cpio.CpioReader
+type newcWriter cpio.CpioWriter
+
+func Header2Info(h *Header, n string, i *cpio.Info) {
```
Comment (hugelgupf) [inline cmds/cpio/pkg/newc/newc.go]: func (h Header) CPIOInfo(...

no need for pointer, as you are not modifying header.

File: cmds/cpio/pkg/types.go [hunk 4]
```diff
@@ -0,0 +1,65 @@
+// Copyright 2013-2017 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// basic types for cpio
+
+package cpio
```
Comment (hugelgupf) [inline cmds/cpio/pkg/types.go]: u-root/pkg/cpio/types.go

File: cmds/cpio/pkg/types.go [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+import (
+	"io"
+)
+
+const (
+	// bad news. These are not defined on
+	// whatever version of Go travis is running.
```
Comment (hugelgupf) [inline cmds/cpio/pkg/types.go]: set go version to 1.9 here? https://github.com/u-root/u-root/blob/master/.travis.yml

**Comment by hugelgupf** [comment]
Go pkgs are usually at the top level. E.g. github.com/u-root/u-root/pkg/cpio/... and then github.com/u-root/u-root/cmds/cpio/... imports that.

I would also give the package a more generic name if *possible*.

File: cmds/cpio/pkg/dat.go [hunk 6]
```diff
@@ -0,0 +1,6 @@
+package cpio
+
+const (
+	Trailer = "TRAILER!!!"
```
Comment (hugelgupf) [inline cmds/cpio/pkg/dat.go]: ?

File: cmds/cpio/pkg/newc/newc.go [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+	t.Pos += int64(len(h))
+	// Make sure it's right.
+	magic := string(h[:6])
+	if magic != newcMagic {
+		return nil, fmt.Errorf("Reader: magic '%s' not a newc file", magic)
+	}
+
+	cpio.Debug("Header is %v\n", h)
+	var hdr Header
+	v := reflect.ValueOf(&hdr)
```
Comment (hugelgupf) [inline cmds/cpio/pkg/newc/newc.go]: Can I send you a pull request to fix all of this?

**Comment by rminnich** [comment]
of course! Thank you!

**Comment by rminnich** [comment]
but:
remember one of the goals. I know the packages should be at top level but we would like our commands to be easily separated as standalone commands.
This structure is like ryan's with fmap/lib

the command can stand alone but provides packages for use by other commands. Go has some of this too as was pointed out.

File: cmds/cpio/pkg/newc/newc.go (hunk 3 above)
Comment (rminnich) [inline cmds/cpio/pkg/newc/newc.go]: Is there an efficiency argument however? Header might get big enough that copying it from stack is not a great thing. I'm not that concerned either way, just a question.

File: cmds/cpio/pkg/types.go (hunk 5 above)
Comment (rminnich) [inline cmds/cpio/pkg/types.go]: It's probably time to do that.

File: cmds/cpio/pkg/dat.go (hunk 6 above)
Comment (rminnich) [inline cmds/cpio/pkg/dat.go]: don't understand the ?

File: cmds/cpio/pkg/newc/newc.go (hunk 7 above)
Comment (rminnich) [inline cmds/cpio/pkg/newc/newc.go]: whatever all of this is, sure.

**Comment by rminnich** [comment]
actually w.r.t. this and discussion with chris, I withdraw my concerns about pkg/ at the top level. When I started this project I don't even think go get existed, or worked. Given that, grabbing a command from u-root is so easy now that we can go ahead with Chris's ideas. Chris kindly offered to take this PR over and that works for me.

**Comment by rminnich** [comment]
Chris, when you take this, let me know.

**Comment by rminnich** [comment]
er, close it.

File: cmds/cpio/pkg/newc/newc.go (hunk 3 above)
Comment (hugelgupf) [inline cmds/cpio/pkg/newc/newc.go]: I would leave that for the optimizer -- that is hopefully basic enough for Go to realize that it shouldn't be passed on the stack.

**Comment by hugelgupf** [comment]
My progress is here https://github.com/hugelgupf/u-root/commits/cpio but it doesn't compile yet. Tomorrow.

**Comment by rminnich** [comment]
superseded! :-)

**Pull request closed without merging**